File: services/gateway/src/util/email.ts
import mjml2html from 'mjml';
import { stripIndent } from 'proper-tags';

const fontFamily = 'Inter, sans-serif';
const color = '#111111';

export function emailHtml(title: string, body: string, minify = true) {
  return mjml2html(
    {
      tagName: 'mjml',
      attributes: {},
      children: [
        {
          tagName: 'mj-head',
          attributes: {},
          children: [
            {
              tagName: 'mj-title',
              attributes: {},
              content: title,
            },
            {
              tagName: 'mj-font',
              attributes: {
                name: 'Inter',
                href: 'https://unpkg.com/@fontsource/inter@5.0.3/index.css',
              },
            },
            {
              tagName: 'mj-preview',
              attributes: {},
              content: body.split('\n\n').at(0) ?? '',
            },
            {
              tagName: 'mj-style',
              attributes: {},
              content: stripIndent`
                a {
                  color: inherit;
                  text-decoration: underline;
                  text-decoration-style: dotted;
                }
                a.plain {
                  text-decoration: none;
                  color: inherit;
                }
                a:hover {
                  text-decoration: underline;
                }
              `,
            },
          ],
        },
        {
          tagName: 'mj-body',
          attributes: {},
          children: [
            {
              tagName: 'mj-section',
              attributes: {},
              children: [
                {
                  tagName: 'mj-column',
                  attributes: {},
                  children: [
                    {
                      tagName: 'mj-image',
                      attributes: {
                        width: '250px',
                        src: 'https://mail.letschurch.cloud/logo.png',
                        href: 'https://lets.church',
                        target: '_blank',
                      },
                    },
                    {
                      tagName: 'mj-text',
                      attributes: {
                        'font-family': fontFamily,
                        'font-size': '16px',
                        color,
                        'line-height': '165%',
                      },
                      content: stripIndent`
                        <h1>${title}</h1>
                        ${body
                          .trim()
                          .split('\n\n')
                          .map((content) => `<p>${content}</p>`)
                          .join('\n')}
                      `,
                      /* children: [ */
                      /*   { */
                      /*     tagName: 'h1', */
                      /*     attributes: {}, */
                      /*     content: title, */
                      /*   }, */
                      /*   ...body.split('\n\n').map((content) => ({ */
                      /*     tagName: 'p', */
                      /*     attributes: {}, */
                      /*     content, */
                      /*   })), */
                      /* ], */
                    },
                  ],
                },
              ],
            },
            {
              tagName: 'mj-section',
              attributes: {},
              children: [
                {
                  tagName: 'mj-column',
                  attributes: {},
                  children: [
                    {
                      tagName: 'mj-social',
                      attributes: {
                        'font-size': '12px',
                        'icon-size': '20px',
                        mode: 'horizontal',
                      },
                      children: [
                        {
                          tagName: 'mj-social-element',
                          attributes: {
                            src: 'https://mail.letschurch.cloud/brand-facebook.png',
                            href: 'https://www.facebook.com/profile.php?id=100092315746719',
                          },
                        },
                        {
                          tagName: 'mj-social-element',
                          attributes: {
                            src: 'https://mail.letschurch.cloud/brand-twitter.png',
                            href: 'https://twitter.com/lets_church',
                          },
                        },
                        {
                          tagName: 'mj-social-element',
                          attributes: {
                            src: 'https://mail.letschurch.cloud/brand-github.png',
                            href: 'https://github.com/LetsChurch',
                          },
                        },
                        {
                          tagName: 'mj-social-element',
                          attributes: {
                            src: 'https://mail.letschurch.cloud/brand-gitlab.png',
                            href: 'https://gitlab.com/LetsChurch',
                          },
                        },
                      ],
                    },
                  ],
                },
              ],
            },
            {
              tagName: 'mj-section',
              attributes: {},
              children: [
                {
                  tagName: 'mj-column',
                  attributes: {},
                  children: [
                    {
                      tagName: 'mj-text',
                      attributes: {
                        'font-family': fontFamily,
                        'font-size': '16px',
                        'font-weight': 'bold',
                        color,
                        align: 'center',
                      },
                      content:
                        '<a href="https://lets.church" class="plain" target="_blank"></a>',
                    },
                    /* { */
                    /*   tagName: 'mj-text', */
                    /*   attributes: { */
                    /*     'font-family', */
                    /*     'font-size': '16px', */
                    /*     'font-weight': 'bold', */
                    /*     color, */
                    /*     align: 'center', */
                    /*   }, */
                    /*   content: 'You received this email because...', */
                    /* }, */
                  ],
                },
              ],
            },
          ],
        },
      ],
    },
    { minify, validationLevel: 'soft' },
  );
}

File: legacy/services/web/cmd/background-worker/util/email.go
package util

import (
	"context"
	"strings"
	"sync"
	"text/template"

	mjml "github.com/Boostport/mjml-go"
	"github.com/samber/oops"
)

const templateText = `
<mjml>
  <mj-head>
    <mj-title>{{.Title}}</mj-title>
    <mj-preview>{{.Preview}}</mj-preview>
    <mj-style>
      a {
				color: inherit;
				text-decoration: underline;
				text-decoration-style: dotted;
      }
      a.plain {
				text-decoration: none;
				color: inherit;
      }
      a:hover {
				text-decoration: underline;
      }
    </mj-style>
  </mj-head>
  <mj-body>
    <mj-section>
      <mj-column>
        <mj-image width="250px" src="https://mail.letschurch.cloud/logo.png" href="https://lets.church" />
        <mj-text font-family="system-ui, sans-serif" font-size="16px" line-height="165%" color="#111111">
          <h1>{{.Title}}</h1>
					{{range .Body}}
					<p>{{.}}</p>
	        {{end}}
        </mj-text>
      </mj-column>
    </mj-section>
    <mj-section>
      <mj-column>
        <mj-social>
          <mj-social-element src="https://mail.letschurch.cloud/brand-facebook.png" href="https://fb.me/LetsChurchOrg" />
          <mj-social-element src="https://mail.letschurch.cloud/brand-x.png" href="https://x.com/LetsChurchOrg" />
          <mj-social-element src="https://mail.letschurch.cloud/brand-github.png" href="https://github.com/LetsChurch" />
          <mj-social-element src="https://mail.letschurch.cloud/brand-gitlab.png" href="https://gitlab.com/LetsChurch" />
        </mj-social>
      </mj-column>
    </mj-section>
  </mj-body>
</mjml>
`

var tmpl = sync.OnceValue(func() *template.Template {
	return template.Must(template.New("email").Parse(templateText))
})

type MakeEmailHtmlArgs struct {
	Title   string
	Preview string
	Body    []string
}

func MakeEmailHtml(args MakeEmailHtmlArgs) (string, error) {
	eb := oops.Hint("MakeEmailHtml")

	var buf strings.Builder
	err := tmpl().Execute(&buf, args)
	if err != nil {
		return "", eb.Hint("Could not compile template to mjml with html/template").Wrap(err)
	}

	output, err := mjml.ToHTML(context.Background(), buf.String(), mjml.WithMinify(true))

	return output, eb.Hint("Could not compile mjml to HTML").Wrap(err)
}
